pkg/sql/schemachanger/scpb/element_collection.go

```diff
@@ -0,0 +1,252 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scpb
+
+import "github.com/cockroachdb/errors"
+
+// ElementCollection represents an ordered set of
+// (current status, target status, element) tuples,
+// looked up in `g` by iterating through `indexes`.
+type ElementCollection[E Element] struct {
+	g       ElementCollectionGetter
+	indexes []int
+}
+
+// ElementCollectionGetter abstracts the actual mechanism for obtaining
+// (current status, target status, element) tuples in an element collection.
+type ElementCollectionGetter interface {
+
+	// Get returns values for the given index in [0, Size()[.
+	Get(index int) (current Status, target TargetStatus, e Element)
+
+	// Size returns the cardinality of the collection.
+	Size() int
+}
+
+// NewElementCollection constructs a generic element collection
+// containing the elements in `g` listed in the `indexes` slice.
+// The slice is owned by the returned struct.
+func NewElementCollection(g ElementCollectionGetter, indexes []int) *ElementCollection[Element] {
+	if g == nil || len(indexes) == 0 {
+		return nil
+	}
+	return &ElementCollection[Element]{
+		g:       g,
+		indexes: indexes,
+	}
+}
+
+// ElementCollection is an ElementCollectionGetter.
+var _ ElementCollectionGetter = &ElementCollection[Element]{}
+
+// Get implements the ElementCollectionGetter interface.
+func (c *ElementCollection[E]) Get(ordinal int) (current Status, target TargetStatus, e Element) {
+	if ordinal < 0 || ordinal >= c.Size() {
+		panic(errors.AssertionFailedf("ordinal %d is out of bounds [0, %d[", ordinal, c.Size()))
+	}
+	return c.g.Get(c.indexes[ordinal])
+}
+
+// Size implements the ElementCollectionGetter interface.
+func (c *ElementCollection[E]) Size() int {
+	if c == nil {
+		return 0
+	}
+	return len(c.indexes)
+}
+
+// Filter generates a new ElementCollection by filtering
+// its values using the predicate.
+func (c *ElementCollection[E]) Filter(
+	predicate func(current Status, target TargetStatus, e E) bool,
+) *ElementCollection[E] {
+	return c.genericFilter(func(current Status, target TargetStatus, genericElement Element) bool {
+		e, ok := genericElement.(E)
+		return ok && predicate(current, target, e)
+	})
+}
+
+// FilterTarget is like Filter but ignoring current status.
+func (c *ElementCollection[E]) FilterTarget(
+	predicate func(target TargetStatus, e E) bool,
+) *ElementCollection[E] {
+	return c.genericFilter(func(current Status, target TargetStatus, genericElement Element) bool {
+		e, ok := genericElement.(E)
+		return ok && predicate(target, e)
+	})
+}
+
+// FilterElement is like Filter but ignoring current and target statuses.
+func (c *ElementCollection[E]) FilterElement(predicate func(e E) bool) *ElementCollection[E] {
+	return c.genericFilter(func(current Status, target TargetStatus, genericElement Element) bool {
+		e, ok := genericElement.(E)
+		return ok && predicate(e)
+	})
+}
```

Comment:
Why would we have those two? Looks like `Filter()` itself is enough; the user can just choose to not use `current` and/or `target` in their supplied predicate.

Reply:
True, but I guess this is just a semantic sugar here, nice to have though.

Reply:
It's syntactic sugar but it's a fair question. I'm anticipating common usage cases here, I identified the common ones by actually using these methods in `scbuildstmt`, but those changes haven't been published yet.

As it turns out we often want to filter based on properties in the element itself.